content/blog/2021-09-22-refactor.md

```diff
@@ -0,0 +1,163 @@
+﻿---
+title: Easy Stuctural Refactors to Python Source Code
+date: 2021-09-22
+description: |
+  Simple, hassle-free, dependency-free, AST based source code refactoring toolkit.
+
+descriptionLong: |
+  [Refactor](https://github.com/isidentical/refactor) is a source code 
+  refactoring engine that takes the advantage of the AST module in the standard
+  library. Through using it, we can find-and-fix patterns in massive codebases.
+
+picture: 2021-09-22/ast.png
+author: batuhan_taskaya
+commentsUrl: https://discuss.dvc.org/t/easy-structural-refactors-to-python-source-code/895
+tags:
+  - Refactor
+  - AST
+---
+
+It is common for bigger codebases to grow to a complexity where it is really
+hard for someone to go and refactor stuff manually on every occurence. There are
+automated solutions for this, like regular expression based search and replace
+tools, but the main problem with those is that they treat source code like a
+regular text document, which might lead to some issues like increased number of
+some false positives (a case where the tool did change when it shouldn't), true
+negatives (a case where the tool didn't change when it should have), or even
+not being able to truly express the structural concepts of the programming
+language in them (since many of the programming languages have actual grammars
+which are impossible to manifest in regexes).
```

Comment:
```suggestion
It is common for big codebases to grow to a complexity where it nearly
impossible for someone to tediously and flawlessly refactor things manually
everywhere. The main problem with existing automated solutions (such as
regex-based find-and-replace tools) is that they treat source code like a plain
text document. This often results in false positives (tools making changes when
they shouldn't), true negatives (not changing what they should). This is
primarily due to a lack of ability to truly encapsulate structural concepts of
the programming language: syntax and grammar that are impossible to manifest in
regexes.
```

Reply:
*is* nearly impossible.  Needs a conjunction (or) in "This often..."  I'll correct and send up with the new tags.  👍 👍 to other changes @casperdcl!  Thank you!

Reply:
I just re-read and edited a few comments, hope you pick up changes.

Reply:
Ok.  @casperdcl should it be and/or?

Reply:
Do these things always happen in conjunction?  (and) It doesn't seem so the way @isidentical wrote it, but I don't understand enough to know.

Reply:
"and" makes more sense ("This [state of affairs] often results in [problem X] and [Y]" rather than "This often results [in problem] [X] or [Y]")

Reply:
> Do these things always happen in conjunction?

identical sentiment to "telling lies leads to unhappiness and/or distrust." Both work; nobody's going to notice difference between `AND` vs `OR` vs `XOR` logic :)